### DISCUSSION TITLE
Feature Request: Read from stream when importing data via Node.JS API

### DISCUSSION THREAD

**Discussion by paulrutter** [duckdb/duckdb#8466] [Ideas]
I tried several things that i found in the documentation (https://duckdb.org/docs/data/json/overview), but i cannot seem to import data directly from a stream in Node.js code.

It would help to import data directly via a Node.js stream instead of a file, to circumvent the disk write overhead.
In my usecase, i'm importing dynamic data into an in-memory duckdb database, query it, and then move to the next part of dynamic data. I want to skip writing the data to disk first, before loading it in duckdb.
Using INSERT statements is also very slow, so using the `read_json_auto` method is preferred.

Desired functionality:
```
const duckdb = require('duckdb');
const { Readable } = require("stream")

const readable = Readable.from([JSON.stringify({
  test: 123
})]);

const db = new duckdb.Database(':memory:');
db.run("CREATE TABLE test1 AS SELECT * FROM read_json_auto('/dev/stdin')", readable, function(err, res) {
	console.error("import done", err);
	db.all('SELECT * from test1', function(err, res) {
	  if (err) {
		throw err;
	  }
	  console.log(res)
	});
});
```

But this fails with:
```
Error: IO Error: No files found that match the pattern "/dev/stdin"] {
  errno: -1,
  code: 'DUCKDB_NODEJS_ERROR',
  errorType: 'IO'
}
```

Am i missing something here? Or is it not possible to import data from a stream directly in Node.js code?

**Comment by mattf96s** [comment]
The Node documentation is a bit confusing in how it shows random incomplete snippets—I would recommend referencing the WASM or Python docs too. Don't forget about the other JSON docs from [here](https://duckdb.org/docs/archive/0.8.1/extensions/json).

If you can, I would query the data directly. Here is an [example](https://duckdb.org/2023/03/03/json.html):
```
SELECT json_group_structure(json)
FROM (
  SELECT *
  FROM read_ndjson_objects('gharchive_gz/*.json.gz')
  LIMIT 2048
);
```
You can also [stream](https://duckdb.org/docs/archive/0.8.1/api/wasm/data_ingestion) the data if it's an Arrow IPC stream and then read it in chunks (from the wasm docs).
```
// ..., from a raw Arrow IPC stream
const streamResponse = await fetch(`someapi`);
const streamReader = streamResponse.body.getReader();
const streamInserts = [];
while (true) {
    const { value, done } = await streamReader.read();
    if (done) break;
    streamInserts.push(c.insertArrowFromIPCStream(value, { name: 'streamed' }));
}
await Promise.all(streamInserts);
```
Another good reference for general DuckDB with Node is the [Mosaic repo](https://github.com/uwdata/mosaic/blob/main/packages/duckdb/src/data-server.js).

**Reply by paulrutter** [reply]
Thanks, i will check those examples out 👍

**Reply by paulrutter** [reply]
@mattf96s I looked into the code snippets, and still haven't been able to load JSON in a streaming way directly in the Node.js API.
There doesn't seem a Node.js equivalent of WASM's `insertArrowFromIPCStream` method, and using WASM in Node.js (although technically possible) seems very convoluted.

The Node.js API does offer a way to stream results back to the client, but no way of streaming data into DuckDB.
The workaround i have now is to create (temporary) JSON files on disk first with the JSON data, which then are copied into DuckDB.
But ideally, one would just stream the data into DuckDB in-process directly.

**Reply by paulrutter** [reply]
Hmm, but i found something that works: https://github.com/duckdb/duckdb/discussions/8680#discussioncomment-6884326.
Need to investigate some more with actual data.

**Comment by paulrutter** [comment]
Related: https://github.com/duckdb/duckdb/discussions/8680

**Comment by paulrutter** [comment]
Relevant test cases that showcase arrow usage in Nodejs: https://github.com/duckdblabs/arrow/blob/main/test/nodejs/arrow_test.js

**Comment by paulrutter** [comment]
The solution as outlined in #8680 is indeed the way to to load JSON data from Node.js into DuckDB.
I used the following approach:

- transform JSON objects into Apache Arrow table
- "insert" (more like reading) the arrow table into DuckDB
```
const duckdb = require('duckdb');
const arrow = require('apache-arrow');

const db = new duckdb.Database(':memory:');
await execute(`INSTALL arrow; LOAD arrow;`); // note; doesn't work on Windows yet

const jsonData = [
  {"userId":1,"id":1,"title":"delectus aut autem","completed":false},
  {"userId":1,"id":2,"title":"quis ut nam facilis et officia qui","completed":false}
];
const arrowTable = arrow.tableFromJSON(jsonData);
await registerArrowTable("table_user", arrowTable);

// SELECT * FROM table_user 

    async function execute(query) {
        return new Promise((resolve, reject) => {
            db.exec(query, (err) => {
                if (err) {
                    return reject(err);
                }
                resolve();
            });
        });
    } 

    async function registerArrowTable(tableName, arrowTable) {
        return new Promise((resolve, reject) => {
            db.register_buffer(tableName, [arrow.tableToIPC(arrowTable)], true, (err, res) => {
                if (err) {
                    return reject(err);
                }
                return resolve(res);
            });
        });
    }
```

The import data performance is very good, as no data needs to be copied.
I might still investigate pushing data "chunks" into DuckDB via Apache Arrow instead of the whole table at once, but this already looks promising and solves the issue i had.

I'm closing this one, as it's solved.

**Reply by tim-at-zenobe** [reply]
Thanks @paulrutter for digging into this and sharing your findings - super useful!

**Reply by selvarajshanmugam** [reply]
@paulrutter Did u find a way to push data chunks from stream into Duckdb table

**Reply by paulrutter** [reply]
No, as in the end I'm using a different tool altogether, as the ingestion speed was faster.

**Reply by selvarajshanmugam** [reply]
Is it possible to tell the ingestion tool details or else if u know any workaround to stream chunk into ducktable

**Reply by paulrutter** [reply]
I ended up using alasql as in memory database. Not as nice as duckdb and less mature, but since data can be queried in memory directly, it was faster for my usecase.
